repo_policy_compliance/blueprint.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provides API blueprint for flask to run the policy checks."""
+
+import logging
+import os
+import secrets
+from typing import cast
+
+from flask import Blueprint, Response, request
+from flask_httpauth import HTTPTokenAuth
+
+from . import Result, all_
+
+repo_policy_compliance = Blueprint("repo_policy_compliance", __name__)
+auth = HTTPTokenAuth(scheme="Bearer")
+runner_tokens: set[str] = set()
+
+# Bandit thinks this is the token value when it is the name of the environment variable with the
+# token value
+CHARM_TOKEN_ENV_NAME = "CHARM_TOKEN"  # nosec
+# Bandit thinks this is the token value when it is the name of the endpoint to get a one time token
+ONE_TIME_TOKEN_ENDPOINT = "/one-time-token"  # nosec
+CHECK_RUN_ENDPOINT = "/check-run"
+CHARM_USER = "charm"
+CHARM_ROLE = CHARM_USER
+RUNNER_USER = "runner"
+RUNNER_ROLE = RUNNER_USER
+
+EXPECTED_KEYS = {
+    "repository_name",
+    "source_repository_name",
+    "target_branch_name",
+    "source_branch_name",
+    "commit_sha",
+}
+
+
+@auth.verify_token
+def verify_token(token: str) -> str | None:
+    """Verify the authentication token.
+
+    Args:
+        token: The token to check.
+
+    Returns:
+        The identity associated with the token or None if no token matches.
+    """
+    charm_token = os.getenv(CHARM_TOKEN_ENV_NAME)
+
+    if not charm_token:
+        logging.error(
+            (
+                "%s environment variable is required for generating one time tokens, it is not "
+                "defined or empty"
+            ),
+            CHARM_TOKEN_ENV_NAME,
+        )
+        return None
+
+    if token == charm_token:
```

Comment:
It would be safer to use `hmac.compare_digest` to compare two secrets to avoid timing attack.

Reply:
Yes that is safer, thanks, fixed